src/main/scala/calc/Typechecker.scala

```diff
@@ -0,0 +1,142 @@
+    case ParamType       => irtpe.AnyType
+    case FunctionType(_) => irtpe.AnyType
+    case NoType          => irtpe.NoType
+  }
+
+  def getType(tree: Tree, scope: TypeScope)
+             (implicit pos: Position): Tree =  tree match {
+    case Literal(x, _)    => Literal(x, DoubleType)
+    case b:      BinaryOp => checkBinop(b, scope)
+    case id:     Ident    => checkIdent(id, scope)
+    case let:    Let      => checkLet(let, scope)
+    case ifExpr: If       => checkIf(ifExpr, scope)
+    case call:   Call     => checkCall(call, scope)
+    case cl:     Closure  => checkClosure(cl, scope)
+  }
+
+  def typecheck(tree: Tree)(implicit pos: Position): Tree = {
+    getType(tree, Map.empty[String, TreeType])
+  }
+
+  private def checkBinop(binop: BinaryOp, scope: TypeScope)
+                        (implicit pos: Position): Tree = {
+    val lhsTyped = getType(binop.lhs, scope)
+    val rhsTyped = getType(binop.rhs, scope)
+    val lType = lhsTyped.tp
+    val rType = rhsTyped.tp
+
+    if (!lType.isFunction && !rType.isFunction)
+      BinaryOp(binop.op, lhsTyped, rhsTyped, lType)
+    else
+      fail(binop.pos, s"Incompatible operand types for ${binop.op}: $lType, $rType")
+  }
+
+  private def checkIdent(ident: Ident, scope: TypeScope)
+                        (implicit pos: Position): Tree =
+    scope.get(ident.name) match {
+      case Some(t) => Ident(ident.name, t)
+      case None    => fail(pos, s"Not in scope: ${ident.name}")
+    }
+
+  private def checkIf(ifExpr: If, scope: TypeScope)
+                     (implicit pos: Position): Tree = ifExpr match {
+    case If(lit@Literal(_, _), thenp, elsep, _) =>
+      val thenTyped = getType(thenp, scope)
+      val elseTyped = getType(elsep, scope)
+      val tType = thenTyped.tp
+      val eType = elseTyped.tp
+
+      if (tType != eType)
+        fail(ifExpr.pos,
+          s"Both branches of if should evaluate to the same type. " +
+          s"Got: $tType and $eType instead.")
+
+      If(getType(lit, scope), thenTyped, elseTyped, tType)
+
+    case _ => fail(ifExpr.pos, s"If condition should be a number")
+  }
+
+  private def checkLet(letExpr: Let, scope: TypeScope)
+                      (implicit pos: Position): Tree =
+    letExpr match {
+      case Let(ident, value, body, _) =>
+        val valueTyped = getType(value, scope)
+        val identTyped = Ident(ident.name, valueTyped.tp)
+        val bodyTyped = getType(body, updateScope(ident.name, valueTyped, scope))
+
+        Let(identTyped, valueTyped, bodyTyped, bodyTyped.tp)
+    }
+
+  private def updateScope(name: String, value: Tree, scope: TypeScope)
+                         (implicit pos: Position): TypeScope = {
+    val valType = getType(value, scope).tp
+    scope + (name -> valType)
+  }
+
+  private def checkCall(call: Call, scope: TypeScope)
+                       (implicit pos: Position): Tree = call match {
+    case Call(fun, args, _) => fun match {
+      case Ident(fname, _) =>
```

Comment:
This sounds much too special-case. What if `fun` is not an `Ident`? Whether the parser can create such a tree or not is pretty much irrelevant. You should be able to handle any kind of `fun`. And if you do, I bet the code will be smaller and more readable.